Handle case-insensitivity in OnDiskCachingFileSystem

benlangmuir opened the pull request:
When encountering a case-insensitive match for a path, attempt to create a fake symlink to the canonical name. This ensures that the realpath and uniqueid of paths will be correct on case-insensitive filesystems.

---

New version of #4223 that only does extra realpath calls only when they are not contained within the original realpath. Also adds test cases involving symlinks to exercise that.

benlangmuir committed:
Refactor FileSystemCache implementations to DiscoveryInstance

benlangmuir committed:
Handle case-insensitivity in OnDiskCachingFileSystem

When encountering a case-insensitive match for a path, attempt to create
a fake symlink to the canonical name. This ensures that the realpath and
uniqueid of paths will be correct on case-insensitive filesystems.

benlangmuir commented:
@swift-ci please test

dexonsmith commented on llvm/lib/CAS/CASFileSystem.cpp:
```diff
@@ -314,34 +314,45 @@ CASFileSystem::getDirectoryIterator(const Twine &Path) {
-      [this](FileSystemCache::DirectoryEntry &Parent,
-             StringRef Name) -> Expected<DirectoryEntry *> {
+namespace {
+class CASFileSystemDI final : public FileSystemCache::DiscoveryInstance {
```
I think `DiscoveryInstanceImpl` would be a better name for this. The whole file is CASFileSystem, so that part isn't as interesting as the `DiscoveryInstance` bit.

dexonsmith commented on llvm/lib/CAS/CachingOnDiskFileSystem.cpp:
```diff
@@ -129,17 +140,76 @@ class CachingOnDiskFileSystemImpl final : public CachingOnDiskFileSystem {
   WorkingDirectoryType WorkingDirectory;
 };
 
+class OnDiskCachingFileSystemDI final
```
Same comment here: I suggest `DiscoveryInstanceImpl`.

dexonsmith commented on llvm/lib/CAS/CASFileSystem.cpp:
```diff
@@ -314,34 +314,45 @@ CASFileSystem::getDirectoryIterator(const Twine &Path) {
+private:
+  using DirectoryEntry = FileSystemCache::DirectoryEntry;
+
+  Expected<DirectoryEntry *> requestDirectoryEntry(DirectoryEntry &Parent,
+                                                   StringRef Name) override {
```
I prefer defining almost everything out-of-line if it's in an anonymous namespace (usually anything more than a one-liner) in order to minimize the extent of the namespace scope. A slight benefit is it makes it easier to add a `static` helper function right next to it, as the implementations evolve. A downside is mentioning the declarations twice.

Up to you what to do here, but thought I'd mention it.

dexonsmith commented on llvm/lib/CAS/CASFileSystem.cpp:
```diff
@@ -314,34 +314,45 @@ CASFileSystem::getDirectoryIterator(const Twine &Path) {
-  };
+  }
+  Error requestSymlinkTarget(DirectoryEntry &Symlink) override {
+    // FIXME: Need to handle lazy symlinks somehow.
```
I'm not sure what a "lazy symlink" even is (I'm sure I wrote the comment, but I wonder if it bitrotted?). If you do, can you move the FIXME to a "header"/`///` comment for `requestSymlinkTarget()` (outside the function) and clarify the text a bit? Otherwise, maybe we can delete it.

benlangmuir replied:
It means nothing to me, I just copied it.  Will remove it.

dexonsmith commented on llvm/lib/CAS/CachingOnDiskFileSystem.cpp:
```diff
@@ -129,17 +140,76 @@ class CachingOnDiskFileSystemImpl final : public CachingOnDiskFileSystem {
+      bool IsTrackingStats, bool LookupOnDisk)
+      : FS(FS), TrackNonRealPathEntries(TrackNonRealPathEntries),
+        IsTrackingStats(IsTrackingStats), LookupOnDisk(LookupOnDisk) {}
+  ~OnDiskCachingFileSystemDI() {}
```
Recently I've been following a style (picked up from some other LLVM people) of putting constructors at the bottom, pretty close to / right next to fields. (My old style was at the top, right after fields.) Up to you though.

dexonsmith commented on llvm/lib/CAS/CachingOnDiskFileSystem.cpp:
```diff
@@ -289,34 +364,86 @@ CachingOnDiskFileSystemImpl::makeFile(DirectoryEntry &Parent,
+  if (KnownStatus) {
+    Status = std::move(*KnownStatus);
+  } else {
+    if (auto EC = sys::fs::status(TreePath, Status, /*follow=*/false))
```
Would this still be clear as `else if`?

dexonsmith commented on llvm/lib/CAS/CachingOnDiskFileSystem.cpp:
```diff
@@ -289,34 +364,86 @@ CachingOnDiskFileSystemImpl::makeFile(DirectoryEntry &Parent,
+  StringRef ParentPath = Parent.getTreePath();
+  std::string Path =
+      (ParentPath.endswith("/") ? ParentPath + Name : ParentPath + "/" + Name)
+          .str();
```
This should have a FIXME about portability / Windows. (I know you're just moving code, but I happened to notice it's missing here.)

dexonsmith commented on llvm/lib/CAS/CachingOnDiskFileSystem.cpp:
```diff
@@ -289,34 +364,86 @@ CachingOnDiskFileSystemImpl::makeFile(DirectoryEntry &Parent,
+    // symlink to an absolute path in the current or a subsequent component, or
+    // that we hit F_GETPATH non-determinism for a hard link on Darwin. We need
+    // to get the realpath of Parent + Name.
+    auto RP = preloadRealPath(Parent, Name);
```
I don't think this call to `preloadRealPath()` will get cached because it's from the `CachingOnDiskFileSystemImpl` instead of the `CachingOnDiskFileSystemDI` (`DiscoveryInstanceImpl`). Should this function take one of the latter as an argument, or maybe be moved to a helper function on it? (or, did I misread the logic? is it okay like this?)

dexonsmith replied:
Oh, I think I did misread it. Can you add "[... to get the realpath of Parent + Name], which is different from the realpath of the file that's ultimately being opened by the calling DiscoveryInstance" (or something to that effect).

I still wonder if this ought to live in DiscoveryInstance; might be more clear that `FS.preload` is being called on purpose, and it's only ever called from there so it might make it more obvious what the context is. Not sure though.

dexonsmith commented on llvm/lib/CAS/CachingOnDiskFileSystem.cpp:
```diff
@@ -64,17 +64,28 @@ class CachingOnDiskFileSystemImpl final : public CachingOnDiskFileSystem {
+                                              DirectoryEntry *KnownRealPath);
+
+  Expected<DirectoryEntry *>
+  lookupOnDiskFrom(DirectoryEntry &Parent, const std::string &TreePath,
```
Why is this not a StringRef?

dexonsmith replied:
Also, wondering if the difference between the name and treepath parameters above is a bit subtle for an overload. Maybe these should have different names?

benlangmuir replied:
> Why is this not a StringRef?

It calls `makeSymlink`, which takes a `const std::string &` in order to use `c_str()` without a copy. Since it's already a string in the callers, I used the same pattern here.  Could use StringRef and copy it; I guess that's what it would do if it were using a normal filesystem API that takes a Twine anyway, this is just an artifact of the fact we're calling readlink directly.

> Maybe these should have different names?

I considered `makeEntryFromStatus` but that's only really accurate when you pass a `KnownStatus`.  Maybe `lookupTreePathOnDisk`?

dexonsmith replied:
> which takes a const std::string & in order to use c_str() without a copy

Aha; that's subtle. Maybe worth a comment?

(Really we should have a `NullTerminatedStringRef : public StringRef` whose type remembers that it's null-terminated, which we can use in situations like this. I've looked at it... a bit hard, but probably worth it.)

dexonsmith replied:
Or I see you committed; seems fine to have the copy. We can fix later if it matters, which it probably doesn't.

dexonsmith replied:
> I considered makeEntryFromStatus but that's only really accurate when you pass a KnownStatus. Maybe lookupTreePathOnDisk?

What you landed looks great.

dexonsmith commented on llvm/lib/CAS/CachingOnDiskFileSystem.cpp:
```diff
@@ -289,34 +364,86 @@ CachingOnDiskFileSystemImpl::makeFile(DirectoryEntry &Parent,
 Expected<FileSystemCache::DirectoryEntry *>
-CachingOnDiskFileSystemImpl::lookupOnDiskFrom(DirectoryEntry &Parent,
-                                              StringRef Name) {
+CachingOnDiskFileSystemImpl::lookupOnDiskFrom(
+    DirectoryEntry &Parent, const std::string &TreePath,
+    Optional<sys::fs::file_status> KnownStatus) {
```
It feels like there should be an assertion:
```
assert(Parent.getTreePath() == dirname(TreePath));
```
If not, return an error if they mismatch?

dexonsmith replied:
(I.e., FileSystemCache has an invariant (not sure if it's enforced?) where the tree path of every DirectoryEntry is the tree path of its parent + "/" + its own name. If we're sure we're respecting that, assertion; if not, error and/or recovery.)

benlangmuir replied:
Good idea. I think we should handle it in `FileSystemCache`, not here. I'll do it separately.

benlangmuir replied:
Looks like it's already asserted in `makeLazyEntry`.

dexonsmith reviewed (approved):
Looks mostly great.
- I was confused by a call to `preloadRealPath()`; I think it's correct, but please double-check.
- A few style suggestions and suggested FIXMEs.

benlangmuir commented:
@swift-ci please test

benlangmuir commented:
@swift-ci please test

benlangmuir commented:
@dexonsmith I took most of your suggestions
* All the style suggestions except the constructor ordering.
* Updated the comment about realpath, and also moved that part of the code into DiscoveryInstance.
* Changed makeSymlink etc. to use StringRef and to copy the path, which behaves more like it would if readlink was in sys::fs.
